### PR TITLE
fix: TypeScript interface overhaul

### PR DESCRIPTION
**Pull request by fbartho**
### Depends On

- https://github.com/infinitered/reactotron-core-client/pull/71

### Fixes

- https://github.com/infinitered/reactotron-react-native/issues/17
- https://github.com/infinitered/reactotron-react-native/issues/60

I think this will work, but the generics are a little magical, so I'm totally looking for feedback if it's too hard to follow!

### CODE REVIEW

**Comment by fbartho** [comment]
Unfortunately, `yarn test` fails on master, so I wasn't able to run the unit tests on this PR

**Comment by rmevans9** [comment]
Hmm... I will check that out. I don't remember tests failing last time I was in here.

**Comment by fbartho** [comment]
@rmevans9 -- there are also typescript failures that are unrelated to my PR, but I think my global typescript is on latest. 

```
💔  {fb/non-breaking-generic-reactotron-for-custom-implementations} ~/Code/reactotron-react-native 🌊  tsc --version
Version 2.8.3
💚  {fb/non-breaking-generic-reactotron-for-custom-implementations} ~/Code/reactotron-react-native 🌊  tsc
node_modules/@types/jest/index.d.ts(385,32): error TS1005: ';' expected.
node_modules/@types/jest/index.d.ts(385,33): error TS1003: Identifier expected.
src/helpers/getReactNativeDimensions.ts(17,4): error TS4060: Return type of exported function has or is using private name 'AppDimensions'.
src/helpers/getReactNativeDimensions.ts(34,53): error TS4060: Return type of exported function has or is using private name 'AppDimensions'.
src/plugins/networking.ts(11,26): error TS4033: Property 'ignoreContentTypes' of exported interface has or is using private name 'RegExp'.
src/plugins/networking.ts(12,18): error TS4033: Property 'ignoreUrls' of exported interface has or is using private name 'RegExp'.
src/plugins/overlay/overlay.tsx(40,43): error TS4020: 'extends' clause of exported class 'FullScreenOverlay' has or is using private name 'Props'.
src/plugins/overlay/overlay.tsx(40,50): error TS4020: 'extends' clause of exported class 'FullScreenOverlay' has or is using private name 'State'.
src/plugins/overlay/overlay.tsx(41,22): error TS4063: Parameter 'props' of constructor from exported class has or is using private name 'Props'.
src/plugins/storybook/storybook.tsx(13,43): error TS4020: 'extends' clause of exported class 'StorybookSwitcher' has or is using private name 'Props'.
src/plugins/storybook/storybook.tsx(13,50): error TS4020: 'extends' clause of exported class 'StorybookSwitcher' has or is using private name 'State'.
💔  {fb/non-breaking-generic-reactotron-for-custom-implementations} ~/Code/reactotron-react-native 🌊  
```

I think these are all warnings, however!

**Comment by rmevans9** [comment]
@fbartho I am trying this out so we can land this today hopefully. I am running into some TS issues when building reactotron-react-native.

```
src/reactotron-react-native.ts:67:7 - error TS2322: Type 'ReactotronImpl' is not assignable to type 'Reactotron<ReactotronReactNative> & ReactotronReactNative'.
  Type 'ReactotronImpl' is not assignable to type 'Reactotron<ReactotronReactNative>'.
    Types of property 'configure' are incompatible.
      Type '(options?: ClientOptions) => Reactotron<ReactotronCore>' is not assignable to type '(options?: ClientOptions) => Reactotron<ReactotronReactNative> & ReactotronReactNative'.
        Type 'Reactotron<ReactotronCore>' is not assignable to type 'Reactotron<ReactotronReactNative> & ReactotronReactNative'.
          Type 'Reactotron<ReactotronCore>' is not assignable to type 'Reactotron<ReactotronReactNative>'.
            Type 'ReactotronCore' is missing the following properties from type 'ReactotronReactNative': useReactNative, overlay, storybookSwitcher

67 const reactotron: Reactotron<ReactotronReactNative> & ReactotronReactNative = createClient(DEFAULTS)
```

This is happening when I link reactotron-react-native to reactotron-core-client with your PR on both ends and try and `yarn build`. Any thoughts?

**Comment by fbartho** [comment]
@rmevans9 I figured out something that was going wrong. -- ReactotronImpl was being exposed in my core client PR, so it was overriding my interface.

Working on it!

**Comment by rmevans9** [comment]
:+1: I am on standby to try it out when you sort it out

**Comment by fbartho** [comment]
@rmevans9 -- figured out my typescript issue. Wasn't running `yarn tsc` and my global was woefully out of date.

Just pushed a patch to reactotron-core-client that makes all the types happy here, want to check?

**Comment by rmevans9** [comment]
Pulled and checking now :+1:
